enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,348 @@
+  a default config file. The default config file could be overriden via a custom component
+  variable.
+
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovnk`, `microshift-storage-topolvm`.
+
+* The `microshift-networking-ovnk` package would include the necessary
+  `/usr/lib/microshift/components.d/networking-ovn/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/storage-topolvm` yamls, including crds.
+
+* The rebase script should be updated to extract those core components
+  as kustomize yamls instead of embedded assets.
+
+* Incompatible component rpms should be impossible to install together.
+  i.e. `microshift-networking-bridge` plus `microshift-networking-ovnk`,
+  this can be handled with Conflict spec clauses.
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the `User histories` section.
+
+#### Upgrading
+
+Upgrading optional components would behave in the same way as upgrading
+components where the assets are embedded into the MicroShift binary.
+
+There is an open question about how to handle migration of CRD based
+content between versions. This problem is not handled or explored in MicroShift
+yet. Could this be handled by one-shot pods, the one-shot pods could be defined to
+be triggered only if an specific type of object/version existed, another
+alternative to this would be to use upgrade-binaries optional in components
+which are ran when necessary (similar to how CNIs work).
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
+when the images are created.
+
+Specific component configuration could be handled as yaml files in
+`/etc/microshift/components.d/` . i.e.:
+
+```
+/etc/microshift/component.d/010-storage-topolvm/config.yaml:
+a:
+  b: 1234
+  c: 5677
+
+y: "abcd"
+```
+
+Would be rendered in the templates with:
+
+```
+{{ .config.a.b || 'default' }} {{ .config.a.c }} {{ config.y }}
+```
+
+Additionally more general MicroShift configuration and system details are passed
+down as variables to the templates:
+
+* `{{ .NodeName }}`
+* `{{ .NodeIP }}`
+* `{{ .ClusterCIDR }}`
+* `{{ .ServiceCIDR }}`
+* `{{ .ClusterDNS }}`
+* `{{ .BaseDomain }}`
+
+#### Deploying Applications
+
+Applications are deployed as usual, we would add support for multiple
+directories in the user component manifests as described in the 2nd goal.
+
+### API Extensions
+
+None.
+
+### Risks and Mitigations
+
+The component combinations would need additional testing handled by
+an increased testing matrix.
+
+Considering networking/storage to be independent elements, probably
+an additional set of tests which cover "microshift-core" alone would
+be enough. When running those tests we need to disable PV and NetworkPolicy
+related testing.
+
```

Comment:
Add a risk that kustomize may not be suitable for handling complex installation or upgrades scenarios?  Similar to the comment from @pmtk about`static assets`. Essentially kustomize is a way for configuration management, not a programmable language to build advanced plugin logic.

With kustomize, it implies that plugin developer would either need to inject the sophisticated logic (validation, installation, upgrade, preparation, cleanup etc) with bash/sh cmds to the container entrypoint script in manifests, or add to the main plugin binary (which may not be desired).

Reply:
I started to think that we should design this in a way that also we could have binary plugins (CNI style), so advanced operations could be performed like that.

I suspect that most of the behavior could go away with go templating + kustomize behavior, but in some areas, we would be limited on what we can do.

But IMHO this binary option would need to be a last resort when we find blocked by the main approach. Because it would create another API to maintain, but we should definitely document and keep the door open for just in case we need it.

Reply:
I will add the risk you are mentioning; I see I forgot to add it on the last pass.

Reply:
Added it now.